Add code coverage reports via Coveralls (#13260)
PowerDNS/pdns: CODE_COVERAGE.md

Code Coverage
-------------

> PowerDNS uses [coveralls](https://coveralls.io/) to generate code coverage reports from our Continuous Integration tests. The resulting analysis can then be consulted [online](https://coveralls.io/github/PowerDNS/pdns), and gives hindsight into which parts of the code are automatically tested.

Reviewer: ```suggestion
PowerDNS uses [coveralls](https://coveralls.io/) to generate code coverage reports from our Continuous Integration tests. The resulting analysis can then be consulted [online](https://coveralls.io/github/PowerDNS/pdns), and gives insight into which parts of the code are automatically tested.
```

> Code coverage is generated during our Continuous Integration tests, for every pull requests and pushs. In addition to the dashboard on coveralls's website, a summary is posted on pull requests.

Reviewer: ```suggestion
Code coverage is generated during our Continuous Integration tests, for every pull request. In addition to the dashboard on Coveralls' website, a summary is posted on pull requests.
```

This approach generates `.gcno` files during the compilation, which are stored along the object files, and `.gcda` files at runtime when the final program is executed.

* There are as many `.gcno` and `.gcda` files as object files, which may be a lot.
* Every invocation of a program updates the `.gcda` files corresponding to the code that has been executed. It will append to existing `.gcda` files, but only process can update a given file so parallel execution will result in corrupted data.
* Writing to each `.gcda` might take a while for large programs, and has been known to slow down execution quite a lot.
* Accurate reporting of lines and branches may be problematic when optimizations are enabled, so it is advised to disable optimizations to get useful analysis.
* Note that the `.gcda` files produced by `clang++` are not fully compatible with the `g++` ones, and with the existing tools, but [`llvm-cov gcov`](https://llvm.org/docs/CommandGuide/llvm-cov.html#llvm-cov-gcov) can produce `.gcov` files that should be compatible. A symptom of this incompatiblity looks like this:

```
Processing pdns/ednssubnet.gcda
__w/pdns/pdns/pdns/ednssubnet.gcno:version '408', prefer 'B02'
```

> ### Sourced Based

Reviewer 2: ```suggestion
### Source Based
```

* Source-based coverage can generate accurate data with optimizations enabled, and has a much lower overhead that `GCOV`.
* The path and exact name of the `.profraw` files generated when a program is executed can be controlled via the `LLVM_PROFILE_FILE` environment variable, which supports [patterns](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#running-the-instrumented-program) like `%p`, which expands to the process ID. That allows running several programs in parallel, each program generating its own file at the end.

## Implementation

We use `clang++`'s source-based coverage method in our CI, as it allows running our regression tests in parallel with several workers. It is enabled by passing the `--enable-coverage=clang` flag during `configure` for all products.

> The code coverage generation is done as part of the (build-and-test-all.yml)[https://github.com/PowerDNS/pdns/blob/master/.github/workflows/build-and-test-all.yml] workflow.

Author: The link is wrong here

* Merge the resulting `*.profraw` file into a single `code.profdata` file by running `llvm-profdata merge -sparse -o /tmp/code.profdata /tmp/code-*.profraw`
* Generate a HTML report into the `/tmp/html-report` directory by running `llvm-cov show --instr-profile /tmp/code.profdata -format html -output-dir /tmp/html-report -object </path/to/product/binary>`

### GCOV

* Run the `configure` script with the `--enable-coverage` option, using either `g++` or `clang++`: `./configure --enable-coverage`
* Compile as usual with: `make`. This will generate `.gcno` files along with the usual `.o` object files and the final binaries.
* Run the test(s) that are expected to cover the new feature, via `./testrunner` or `make check` for the unit tests, and the instructions of the corresponding `regression-tests*` directory for the regression tests. Note that the regression should not be run in parallel, as it would corrupt the `.gcna` files that will be generated in the process. For dnsdist, that means running `pytest` without the `--dist=loadfile -n auto` options.
* Generate a HTML report using `gcovr`, or `gcov` then `lcov`

# Remaining Tasks

> The way our code coverage report is generated does not currently handle very well the multiple tools that are generated during the authoritative server build, and end up in the `pdns-tools` package. Consequently the coverage report for these tools, and the related code parts, is not accurate.

Reviewer: ```suggestion
The way our code coverage report is generated does not currently handle the different authoritative server tools (that end up in the `pdns-tools` package) very well. Consequently the coverage report for these tools, and the related code parts, is not accurate.
```
